File: iris-vc/src/main/java/city/smartb/iris/vc/VerifiableCredential.java
package city.smartb.iris.vc;

import city.smartb.iris.jsonld.JsonLdObject;
import city.smartb.iris.jsonld.reader.JsonField;
import city.smartb.iris.jsonld.reader.JsonFieldReader;
import city.smartb.iris.ldproof.LdProof;

import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

public class VerifiableCredential extends JsonLdObject {

    public static VerifiableCredential from(LinkedHashMap<String, Object> json) {
        return new VerifiableCredential(json);
    }

    public static final String VC_ISSUER = "issuer";
    public static final String VC_ISSUANCE_DATE = "issuanceDate";
    public static final String VC_CREDENTIAL_SUBJECT = "credentialSubject";

    public VerifiableCredential(Map<String, Object> json) {
        super(json);
    }

    public VerifiableCredential(Map<String, Object> json, JsonFieldReader reader) {
        super(json, reader);
    }

    public String getIssuanceDate() {
        return this.get(VC_ISSUANCE_DATE).asString();
    }

    public String getIssuer() {
        return this.get(VC_ISSUER).asString();
    }

    public <T> T getCredentialSubject(Class<T> clazz) {
        return this.get(VC_CREDENTIAL_SUBJECT).asObject(clazz);
    }

    public <T> List<T> getCredentialSubjectAsList(Class<T> clazz) {
        return this.get(VC_CREDENTIAL_SUBJECT).asListObjects(clazz);
    }

    public <T> JsonField getCredentialSubject() {
        return this.get(VC_CREDENTIAL_SUBJECT);
    }

    public LdProof getProof() {
        Map<String, Object> json = this.get(LdProof.JSON_LD_PROOF).asMap();
        return LdProof.fromMap(json);
    }

}

File: iris-ld/iris-vc/src/main/kotlin/city/smartb/iris/ld/vc/VerifiableCredential.kt
package city.smartb.iris.ld.vc

import city.smartb.iris.ld.jsonld.JsonLdObject
import city.smartb.iris.ld.jsonld.reader.JsonField
import city.smartb.iris.ld.jsonld.reader.JsonFieldReader
import city.smartb.iris.ld.ldproof.LdProof
import com.fasterxml.jackson.databind.annotation.JsonDeserialize
import vc.deserializer.VerifiableCredentialDeserializer

@JsonDeserialize(using = VerifiableCredentialDeserializer::class)
class VerifiableCredential : JsonLdObject {
    constructor(json: Map<String, Any>) : super(json)
    constructor(json: Map<String, Any>, reader: JsonFieldReader) : super(json, reader)


    companion object {
        fun from(json: LinkedHashMap<String, Any>): VerifiableCredential {
            return VerifiableCredential(json)
        }
        const val VC_ISSUER = "issuer"
        const val VC_ISSUANCE_DATE = "issuanceDate"
        const val VC_CREDENTIAL_SUBJECT = "credentialSubject"
    }

    val issuanceDate: String?
        get() = this.get(VC_ISSUANCE_DATE).asString()
    val issuer: String?
        get() = this.get(VC_ISSUER).asString()

    fun <T> getCredentialSubject(clazz: Class<T>?): T {
        return this.get(VC_CREDENTIAL_SUBJECT).asObject(clazz)
    }

    fun <T> getCredentialSubjectAsList(clazz: Class<T>?): List<T> {
        return this.get(VC_CREDENTIAL_SUBJECT).asListObjects(clazz)
    }

    fun <T> getCredentialSubject(): JsonField {
        return this.get(VC_CREDENTIAL_SUBJECT)
    }

    val proof: LdProof
        get() {
            val json: Map<String, Any> = this.get(LdProof.JSON_LD_PROOF).asMap()
            return LdProof.fromMap(json)
        }
}
